nibabel/deprecated.py

```diff
@@ -63,3 +77,121 @@ def __init__(self, *args, **kwargs):
+
+    Ensures that `text` always ends with a carriage return
+    """
+    return text.rstrip() + '\n'
+
+
+def _add_dep_doc(old_doc, dep_doc):
+    """ Add deprecation message `dep_doc` to docstring in `old_doc`
+
+    Parameters
+    ----------
+    old_doc : str
+        Docstring from some object.
+    dep_doc : str
+        Deprecation warning to add to top of docstring, after initial line.
+
+    Returns
+    -------
+    new_doc : str
+        `old_doc` with `dep_doc` inserted after any first lines of docstring.
+    """
+    if not old_doc:
+        return _ensure_cr(dep_doc)
+    old_doc = _ensure_cr(old_doc)
+    dep_doc = _ensure_cr(dep_doc)
+    old_lines = old_doc.splitlines()
+    new_lines = []
+    for line_no, line in enumerate(old_lines):
+        if line.strip():
+            new_lines.append(line)
+        else:
+            break
+    next_line = line_no + 1
+    if next_line >= len(old_lines):
+        # nothing following first paragraph, just append message
+        return old_doc + '\n' + dep_doc
+    indent = _LEADING_WHITE.match(old_lines[next_line]).group()
+    dep_lines = [indent + L for L in [''] + dep_doc.splitlines() + ['']]
+    return '\n'.join(new_lines + dep_lines + old_lines[next_line:]) + '\n'
+
+
+def deprecate_with_version(message, since='', until='',
+                           warn_class=DeprecationWarning,
+                           expired_class=ExpiredDeprecationError):
+    """ Decorator that marks function or method as deprecated
+
+    The decorated function / method will:
+
+    * Raise the given `warning_class` warning when the function / method gets
+      called, up to (and including) version `until` (if specified);
+    * Raise the given `expired_class` error when the function / method gets
+      called, when the package version is greater than version `until` (if
+      specified).
+
+    Parameters
+    ----------
+    message : str
+        Message explaining deprecation, giving possible alternatives.
+    since : str, optional
+        Released version at which object was first deprecated.
+    until : str, optional
+        Last released version at which this function will still raise a
+        deprecation warning.  Versions higher than this will raise an error.
+    warn_class : class, optional
+        Class of warning to generate.
+    """
+    messages = [message]
+    if (since, until) != ('', ''):
+        messages.append('')
+    if since:
+        messages.append('* deprecated from version: ' + since)
+    if until:
+        expired = PKG_VERSION > LooseVersion(until)
+        messages.append('* {} a {} as of version: {}'.format(
+            "Raises" if expired else "Will raise",
+            expired_class,
+            until))
+    message = '\n'.join(messages)
+
+    def deprecator(func):
+
+        @functools.wraps(func)
+        def deprecated_func(*args, **kwargs):
+            if expired:
+                raise expired_class(message)
```

Comment:
My worry is that someone (presumably me) will forget to remove them in tim, and some code will get out into the wild in that state.   In that case, then the error will only occur if they try and use the functions that they shouldn't be using, instead of completely breaking the module import.

I think all these functions have their deprecations checked, so they should generate a test error when they expire.